exercise-book/src/iterators.md

```diff
@@ -0,0 +1,324 @@
+- You can parallelize your code with minimal changes using [rayon](https://crates.io/crates/rayon).
+
+The first point is not in vain - the original snippet has a bug in the upper bound, since `0..=v.len()` is inclusive!
+
+Think of iterators as lazy functions - they only carry out computation when called with a `.collect()` or similar, not the `.map()` itself.
+
+### Turbo fish syntax `::<>`
+
+Iterators sometimes struggle to figure out the types of all intermediate steps and need assistance.
+
+We can write
+
+```rust
+let xs = v.iter()
+          .map(|elem| elem * 2)
+          .collect::<Vec<i32>>();
+```
+
+instead to avoid having a `xs: Vec<_> = ...`. This `::<SomeType>` syntax is called the [turbo fish operator](https://doc.rust-lang.org/book/appendix-02-operators.html?highlight=turbo%20fish#non-operator-symbols), and it disambiguates calling the same method with different output types, like `.collect::<HashSet<i32>>()` and `.collect::<Vec<i32>>()` (try it!)
+
+### Dereferences
+
+Rust will often admonish you to add an extra dereference (`*`) by comparing the expected input and actual types, and you'll need to write something like `.map(|elem| *elem * 2)` to correct your code. A tell tale sign of this is that the expected types and the actual type differ by the number of `&`'s present.
+
+Remember you can select and hover over each expression and rust-analyzer will display its type if you want a more detailed look inside.
+
+## Destructuring in closures
+
+Not all iterator chains operate on a single iterable at a time. This may mean joining several iterators and processing them together by destructuring a tuple when declaring the closure:
+
+```rust
+let x = [10, 20, 30];
+let y = [1, 2, 3];
+let z = x.iter().zip(y.iter())
+         .map(|(a, b)| a * b)
+         .sum::<i32>();
+```
+
+where the `.map(|(a, b)| a + b)` is iterating over `[(10, 1), (20, 2), (30, 3)]` and calling the left argument `a` and the right argument `b`, in each iteration.
+
+## Step-by-Step-Solution
+
+⚠️ NOTICE! ⚠️
+
+When starting out with iterators, it's very easy to be "led astray" by doing what is locally useful as suggested by the compiler.
+
+Concretely, our first solution will feel like a slog because we'll deal with a lot of `Option` and `Result` wrapping and unwrapping that other languages wouldn't make explicit.
+
+A second more idiomatic solution will emerge in `Step 6` once we learn a few key idioms from the standard library.
+
+You, unfortunately, relive similar experiences when learning Rust without knowing the right tools from the standard library to handle errors elegantly.
+
+🧘 END OF NOTICE 🧘
+
+We highly recommend that you consider turning off `inlay hints` in your `rust-analyzer` settings to `offUnlessPressed`, as they can get very noisy very quickly. You can do this by searching for `inlay hints` and choosing the right option in `Settings > Editor > Inlay Hints > Enabled`.
```

Comment:
Disagree. Inlay hints show the arguments in `map` and `filter` calls and help people figure out problems with extra `&` in there. For a numeric example we have here the hints are very helpful.

A better advice would be to:

1. Start every iterator call on a new line, so that you can see closure arguments and type hints for the iterator at the end of the line clearly.
2. When in doubt, write `.map(|x| x)` first to see what item types you get and decide on what iterator methods to use and what to do inside a closure based on that.

Reply:
Deleted the last line and moved it to a new section: `## Iterator chains workflow advice`.